_posts/2022-08-22-fan-out-and-back-in-using-functions.md

```diff
@@ -0,0 +1,415 @@
+
+The following code creates a connection to Redis using redis-py:
+
+```python
+redisClient = None
+
+def initRedis():
+    redisHostname = os.getenv('redis_hostname')
+    redisPort = os.getenv('redis_port')
+
+    with open('/var/openfaas/secrets/redis-password', 'r') as s:
+        redisPassword = s.read()
+
+    return redis.Redis(
+        host=redisHostname,
+        port=redisPort,
+        password=redisPassword,
+    )
+```
+
+Initialize the redis client in the function handler:
+
+```python
+def handle(event, context):
+    global redisClient
+
+    if redisClient == None:
+        redisClient = initRedis()
+```
+
+Update the stack.yaml for the functions to add the Redis environment variables and secret: 
+```yaml
+environment:
+    s3_bucket: of-demo-inception-data
+    redis_hostname: "redis-master.redis.svc.cluster.local"
+    redis_port: 6379
+secrets:
+    - s3-key
+    - s3-secret
+    - redis-password
+```
+
+Don't forget to add `redis` to the `requirements.txt` for your functions.
+
+The `create-batch` function can now be updated to count the number of URLs in the CSV file and initialize a Redis key for `batchId` with that value. 
+```python
+batchFile = event.body.decode()
+s3URL = 's3://{}/{}'.format(bucketName, batchFile)
+with open(s3URL, 'rb', transport_params={'client': session.client('s3')}) as f:
+    records = pd.read_csv(f)
+
+batchId = str(uuid.uuid4())
+batchSize = len(records)
+
+# Initialize a counter in Redis
+redisClient.set(batchId, batchSize)
+
+for index, col in records.iterrows():
+    headers = { 'X-Batch-Id': batchId }
+    res = requests.post('http://gateway.openfaas:8080/async-function/run-model', data=col['url'], headers=headers)
+```
+
+In the `run-model` function we need to decrement this counter every time an invocation to the `inception` function finishes. If the counter reaches zero, this means all work in the batch is completed and the final function can be invoked.
+
+```python
+
+remainingWork = redisClient.decr(batchId)
+
+if remainingWork == 0:
+    headers = { 'X-Batch-Id': batchId }
+    res = requests.post("http://gateway.openfaas:8080/async-function/collect-result", headers=headers)
+    r.delete(batchId)
+```
+
+### Collect the results
+The final function is called when all asynchronous requests in a batch are completed. This function can be used to aggregate the results, notify some other actor or trigger the next step in your processing workflow. In this example we are going to use it to collect all the responses in a single json file and upload that file to the S3 bucket.
+
+The code of the `collect-result` function:
+
+```python
+import os
+import json
+import boto3
+
+from smart_open import s3
+from smart_open import open
+
+def handle(event, context):
+    bucketName = os.getenv('s3_bucket')
+
+    with open('/var/openfaas/secrets/s3-key', 'r') as s:
```

Comment:
Could those steps move into the outer scope? They won't change so could be run only once.